hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/schemas/InitialModServiceTokenSchema.java

```diff
@@ -184,8 +190,47 @@ public void setStakingFs(
         this.mnc = mnc;
     }
 
+    /**
+     * Updates in-state staking info to match the address book.
+     * <ol>
+     *     <li>For any node with staking info in state that is no longer in the address book,
+     *     marks it deleted and sets its weight to zero.</li>
+     *     <li>For any node in the address book that is not in state, initializes its staking info.</li>
+     * </ol>
+     *
+     * @param ctx {@link MigrationContext} for this schema restart operation
+     */
     @Override
-    public void restart(@NonNull MigrationContext ctx) {}
+    public void restart(@NonNull MigrationContext ctx) {
+        final var networkInfo = ctx.networkInfo();
+        final var newStakingInfos = ctx.newStates().<EntityNumber, StakingNodeInfo>get(STAKING_INFO_KEY);
+        // We need to validate and mark any node that are removed during upgrade as deleted.
+        // Since restart is called in the schema after an upgrade, and we don't want to depend on
+        // schema version change, validate all the nodeIds from the addressBook in state and mark
+        // them as deleted if they are not yet deleted in staking info.
+        if (!ctx.previousStates().isEmpty()) {
+            final var oldStakingInfos = ctx.previousStates().<EntityNumber, StakingNodeInfo>get(STAKING_INFO_KEY);
+            oldStakingInfos.keys().forEachRemaining(nodeId -> {
+                final var stakingInfo = requireNonNull(oldStakingInfos.get(nodeId));
+                if (!networkInfo.containsNode(nodeId.number()) && !stakingInfo.deleted()) {
+                    newStakingInfos.put(
+                            nodeId,
+                            stakingInfo.copyBuilder().weight(0).deleted(true).build());
+                    log.info(
+                            "Node {} is marked deleted since it is deleted from addressBook during restart.",
+                            nodeId.number());
+                }
+            });
+        }
+        // Validate if any new nodes are added in addressBook and not in staking info.
+        // If so, add them to staking info/ with weight 0. Also update maxStake and
+        // minStake for the new nodes.
+        addAdditionalNodesIfAny(newStakingInfos, networkInfo.addressBook(), ctx.configuration());
+
+        if (newStakingInfos.isModified()) {
+            ((WritableKVStateBase<EntityNumber, StakingNodeInfo>) newStakingInfos).commit();
```

Comment:
This looks like a smell. We should never commit in a service.

Do you think this is even necessary? In `MerkleSchemaRegistry`, we call `commit()` right after calling `restart()` anyway.

Reply:
Agreed, should be unnecessary, removed.